# Quality control and filtering

## Learning objectives

## Overview

* Basic QC plots + filtering (20min)
* (Negative probes if using cosmx)
* Identify low quality cells

```{r}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE
)

# load libraries for this step
library(Seurat)
library(tidyverse)
library(here)
```

Read in subsampled seurat object

## Identifying potential technical artefacts

```{r}
# Read in the subsampled data for this lesson
so_path <- here("data", "GSE234713_CosMx_IBD_seurat_00_raw_subsampled.RDS")
so <- readRDS(so_path)
```

Library size (total transcripts) and cell areas are good indicators of technical
artefacts introduced during data generation steps. 

Possible artefacts can include:

- Low quality cells  
- Cell segmentation challenges (combining two cells, not aligned with real cell
boundaries, leading to high cell area and low RNA count)  

Understanding how these are prevalent in your data can help with biological
interpretation downstream. For example, discerning whether differences in cells
are due to e.g. cell type differences, or technical artefacts.

### Library size (transcript counts)

```{r warning=F}
ggplot(so@meta.data, aes(x = nCount_RNA, col = tissue_sample)) +
  geom_density() +
  scale_x_log10() +
  theme_bw() +
  annotation_logticks(sides = "b", colour = "lightgrey") +
  ggtitle("Transcript counts per cell")
```

The transcripts per cell all display a log-normal distribution with shifting medians.

Indicates no largely low-quality cells, and a good example of library size 
effects that should be normalised downstream, prior to analyses.

### Cell surface area

```{r warning=F}
# Total counts per cell
ggplot(so@meta.data, aes(x=Area, col=orig.ident)) +
  geom_density() +
  scale_x_log10() +
  theme_bw() +
  annotation_logticks(sides = "b") +
  ggtitle("Area (px) per cell")
```

Areas also follow log-normal distributions with medians ranging between 3,000 -
4,000px^2. Sample/FOV `CD.c` has slightly smaller cell area with a median
of ~2,000px^2.

## Bivariate QC

Useful to view bivariate metrics - the larger the cell, the more transcripts, in
general.

```{r, warning=F}
ggplot(so@meta.data, aes(x=Area, y=nCount_RNA, colour=tissue_sample)) +
  geom_point(size = 0.1, alpha = 0.4) +
  scale_x_log10() +
  scale_y_log10() +
  annotation_logticks(sides = "bl") +
  facet_wrap(~orig.ident) +
  annotation_logticks(colour = "lightgrey") +
  theme_light()
```

Another good check for low-quality cells is to view the bivariate distribution
of transcript counts per cell vs. how many of these have been mapped successfully
to a gene.

```{r}
so@meta.data %>%
  ggplot(aes(x = nCount_RNA, y = nFeature_RNA, colour = tissue_sample)) +
  geom_point(alpha = 0.4, size = 0.1) +
  facet_wrap(~tissue_sample) +
  theme_light() +
  theme(legend.position = "none")
```

The more transcripts detected `nCount_RNA`, the more mapped genes there should be.

::: {.callout-note title="Cell size variability"}
Cautious of samples with highly variable varying sizes. Total transcript
counts and cell areas can vary substantially. Applying "standard" QC or
filtering thresholds uniformly across all cells may remove valid cells.

![](images/hatton2023_fig3.jpg){width=90%}
:::

Point out where bad quality cells would reside

## Filtering low-quality cells

### Negative probes as an indicator of noise  

Spatial transcriptomics assays include both RNA-specific probes, and negative
control probes (or, background probes). While RNA specific-probes hybridise to
specific targets, negative probes do not bind to anything. Their primary purpose
is to quantify and account for this technical noise and non-specific signal.

In this step, we will calculate the proportion of negative control probes
relative to the total counts (RNA) for each cell. Cells exhibiting a high
percentage of noise (i.e., high proportion of negative probe counts) can be
flagged and removed during the QC stage.

Key metrics to add to the Seurat object:

- `so$percent_neg`: The percentage of negative probes per cell
- `so$mean_neg`: The average negative probes per cell

```{r}
so$percent_neg <- (so$nCount_negprobes / (so$nCount_RNA + so$nCount_negprobes)) * 100

# Join layers to calculate average_neg
so@assays$negprobes <- JoinLayers(so@assays$negprobes)
so$mean_neg <- colMeans(so@assays$negprobes) # Only defined first sample
```

```{r}
so@meta.data %>%
  ggplot(aes(x = percent_neg, colour = tissue_sample)) +
  geom_density() +
  scale_x_log10() +
  theme_light() +
  annotation_logticks(sides = "b", colour = "lightgrey")
```

Log-normally distributed. CD_b, HC_B samples are less "noisy".

```{r}
so@meta.data %>%
  ggplot(aes(x = tissue_sample, y = mean_neg, colour = tissue_sample)) +
  geom_col() +
  theme_light()
```

```{r}
# Discuss difference with Xenium - xenium lower counts, lower background.
ggplot(so@meta.data, aes(y=mean_neg, x=nCount_RNA)) +
  geom_point(pch=3, alpha=0.1) +
  scale_x_log10() +
  theme_bw() +
  ggtitle("Negative probes vs counts")

ggplot(so@meta.data, aes(y=percent_neg, x=nCount_RNA)) +
  geom_point(pch=3, alpha=0.1) +
  scale_x_log10() +
  coord_cartesian(ylim = c(0, 10)) +
  facet_wrap(~tissue_sample) +
  theme_bw() +
  ggtitle("Negative probes vs counts")
```

```{r}
min_ncount_rna <- 90
max_percent_neg <- 2
```

```{r}
ggplot(so@meta.data, aes(y=percent_neg, x=nCount_RNA)) +
  geom_point(pch=3, alpha=0.1) +
  geom_hline(yintercept = max_percent_neg, lty=3, colour = "red") +
  geom_vline(xintercept = min_ncount_rna, lty=3, colour = "red") +
  scale_x_log10() +
  coord_cartesian(ylim = c(0, 21)) +
  facet_wrap(~tissue_sample) +
  theme_bw() +
  ggtitle("Negative probes vs counts")
```

### Identifying filtering thresholds

```{r}
# Apply a filter
so@meta.data <- 
  so@meta.data %>%
  mutate(
    qc = if_else(nCount_RNA >= min_ncount_rna & percent_neg <= max_percent_neg, "Keep", "Remove")
  )

table(so$qc, so$orig.ident)
```

Nearly 50% of the cells for HC.c will be filtered out, display which cells will
be removed.

```{r}
ImageDimPlot(so, fov = "GSM7473684.HC.c", group.by = "qc")
```

The stromal compartment will be removed entirely and can be difficult for
co-localisation analyses. Further, the epithelial compartment in the bottom left
FOV is very patchy.

## Exercise: Identifying your own filtering thresholds

You have seen an example where stringent filtering omits key cells or
compartments.

Identify a suitable threshold for `min_ncount_rna` and `max_percent_neg` by
iteratively:

1. Changing the thresholds for `min_ncount_rna` and `max_percent_neg`
2. Displaying the bivariate distribution of `nCount_RNA` and `percent_neg` per
tissue sample
3. Using `ImageDimPlot` to visualise which cells are removed

```{r}
min_ncount_rna <- 70
max_percent_neg <- 4
```

```{r}
ggplot(so@meta.data, aes(y=percent_neg, x=nCount_RNA)) +
  geom_point(pch=3, alpha=0.1) +
  geom_hline(yintercept = max_percent_neg, lty=3, colour = "red") +
  geom_vline(xintercept = min_ncount_rna, lty=3, colour = "red") +
  scale_x_log10() +
  coord_cartesian(ylim = c(0, 21)) +
  facet_wrap(~tissue_sample) +
  theme_bw() +
  ggtitle("Negative probes vs counts")
```

```{r}
so@meta.data <- 
  so@meta.data %>%
  mutate(
    qc = if_else(nCount_RNA >= min_ncount_rna & percent_neg <= max_percent_neg, "Keep", "Remove")
  )

table(so$qc, so$orig.ident)
```

```{r}
# Epithelial compartment less patchy, but still hard to retain the stromal cells
# These are the trade-offs to consider
ImageDimPlot(so, fov = "GSM7473684.HC.c", group.by = "qc")
```

```{r}
# A lot of cells marked for removal and a bit patchy
ImageDimPlot(so, fov = "GSM7473688.CD.a", group.by = "qc")
```

### Applying the filter

```{r}
so_filtered <- subset(so, qc == "Keep")
table(so_filtered$orig.ident)
```

```{r eval=F}
saveRDS(so_filtered, file = here("data", "GSE234713_CosMx_IBD_seurat_02_rna70_neg4.RDS"))
```

## Field of View QC (CosMX only; Advanced)

Good to check library sizes across FOVs - red flags include FOVs with close to
no cells.

```{r}
so@meta.data %>%
  unite(col = unique_fov_names, orig.ident, fov, remove = F) %>%
  ggplot(aes(x = nCount_RNA, col = orig.ident, group = unique_fov_names)) +
  geom_density() +
  scale_x_log10() +
  theme_light() +
  annotation_logticks(sides = "b")
```

TODO: As well as border effects. Run per sample.

```{r, eval=F}
# For each cell, calculate the distance between the centroid to each FOV border
so_borders <- so@meta.data %>%
  group_by(tissue_sample, fov) %>%
  mutate(
    top = max(CenterY_local_px) - CenterY_local_px,
    bottom = CenterY_local_px - min(CenterY_local_px),
    left = CenterX_local_px - min(CenterX_local_px),
    right = max(CenterX_local_px) - CenterX_local_px
  ) %>%
  select(top, bottom, left, right, nCount_RNA, tissue_sample) %>%
  pivot_longer(cols = -c(nCount_RNA, tissue_sample), names_to = "fov_direction", values_to = "distance") 

so_borders %>%
  ggplot(aes(x = distance, y = nCount_RNA)) +
  geom_point(size = 0.1, alpha = 0.4) +
  facet_wrap(~fov_direction, ncol = 4) +
  scale_y_log10() +
  theme_light() +
  geom_smooth(method = "loess", span = 0.2)
```

## Summary

Quality control (QC) in spatial transcriptomics is a highly iterative and
context-dependent process. It requires continual exploration of your data rather
than following a prescriptive set of rules. When examining and filtering your
data for the first time, we recommend setting permissive threshold so
meaningful cells are not removed. After a preliminary look at the data, QC steps
can be revisted.

In comparison to older transcriptomics technologies, spatial technologies introduce
new sources of technical variation that can make interpretation of your results
tricky, and even obscure true, biological signals in the data. While you do not
need to filter in every step, it is useful to understand these potential sources
of bias, and inform subsequent rounds of QC.

Sources of technical variation are also platform dependent and should be
accounted for. For example, CosMx contains FOVs which adds another dimension,
compared to Xenium.

Quality control is not a step you conduct once during a pre-processing pipeline.
Throughout the following steps in the workshop, you will continue to conduct QC
steps.
