<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 12 Clustering Labeling | Intro to Spatial Transcriptomics</title>
<meta name="author" content="">
<meta name="description" content="Classifing cells into meaningful celltypes (or cell states) is a time consuming, but extremely improtant, part of a spatial analysis. It may involve pulling together multiple lines of evidence to...">
<meta name="generator" content="bookdown 0.42.1 with bs4_book()">
<meta property="og:title" content="Chapter 12 Clustering Labeling | Intro to Spatial Transcriptomics">
<meta property="og:type" content="book">
<meta property="og:description" content="Classifing cells into meaningful celltypes (or cell states) is a time consuming, but extremely improtant, part of a spatial analysis. It may involve pulling together multiple lines of evidence to...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 12 Clustering Labeling | Intro to Spatial Transcriptomics">
<meta name="twitter:description" content="Classifing cells into meaningful celltypes (or cell states) is a time consuming, but extremely improtant, part of a spatial analysis. It may involve pulling together multiple lines of evidence to...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.7.0/transition.js"></script><script src="libs/bs3compat-0.7.0/tabs.js"></script><script src="libs/bs3compat-0.7.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.34.0/datatables.js"></script><link href="libs/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.1/js/crosstalk.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Intro to Spatial Transcriptomics</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> Overview</a></li>
<li><a class="" href="intro.html"><span class="header-section-number">2</span> Introduction</a></li>
<li><a class="" href="background.html"><span class="header-section-number">3</span> Background</a></li>
<li><a class="" href="exampleexperiment.html"><span class="header-section-number">4</span> Example Experiment</a></li>
<li><a class="" href="aims.html"><span class="header-section-number">5</span> Aims for today</a></li>
<li><a class="" href="input-data.html"><span class="header-section-number">6</span> Input Data</a></li>
<li><a class="" href="qc.html"><span class="header-section-number">7</span> QC</a></li>
<li><a class="" href="basic-visualisation.html"><span class="header-section-number">8</span> Basic visualisation</a></li>
<li><a class="" href="reduce-dimensionality---umaps.html"><span class="header-section-number">9</span> Reduce dimensionality - UMAPs</a></li>
<li><a class="" href="batch-correction.html"><span class="header-section-number">10</span> Batch Correction</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">11</span> Clustering</a></li>
<li><a class="active" href="clustering-labeling.html"><span class="header-section-number">12</span> Clustering Labeling</a></li>
<li><a class="" href="analysis-example.html"><span class="header-section-number">13</span> Analysis example</a></li>
<li><a class="" href="analysis-example-1.html"><span class="header-section-number">14</span> Analysis example</a></li>
<li><a class="" href="run-moransi-on-all-samples.html"><span class="header-section-number">15</span> Run MoransI on all samples</a></li>
<li><a class="" href="next-steps.html"><span class="header-section-number">16</span> Next steps</a></li>
<li><a class="" href="resources.html"><span class="header-section-number">17</span> Resources</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clustering-labeling" class="section level1" number="12">
<h1>
<span class="header-section-number">12</span> Clustering Labeling<a class="anchor" aria-label="anchor" href="#clustering-labeling"><i class="fas fa-link"></i></a>
</h1>
<p>Classifing cells into meaningful celltypes (or cell states) is a time consuming, but extremely improtant, part of a spatial analysis.</p>
<p>It may involve pulling together multiple lines of evidence to assign celltype labels to cluster labels. Some approaches outlined below:</p>
<div id="cluster-markers" class="section level2" number="12.1">
<h2>
<span class="header-section-number">12.1</span> Cluster Markers<a class="anchor" aria-label="anchor" href="#cluster-markers"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">so</span>, group.by<span class="op">=</span><span class="st">'seurat_clusters'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-37-1.png" width="672"></div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/SplitLayers.html">JoinLayers</a></span><span class="op">(</span><span class="va">so</span><span class="op">)</span></span>
<span><span class="va">marker_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">so</span>, group.by<span class="op">=</span><span class="st">'seurat_clusters'</span>, only.pos<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
## ℹ Please use the `layer` argument instead.
## ℹ The deprecated feature was likely used in the Seurat package.
##   Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.</code></pre>
<pre><code>## Warning: `PackageCheck()` was deprecated in SeuratObject 5.0.0.
## ℹ Please use `rlang::check_installed()` instead.
## ℹ The deprecated feature was likely used in the Seurat package.
##   Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.
## This warning is displayed once every 8 hours.
## Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.</code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marker_table_top</span> <span class="op">&lt;-</span> <span class="va">marker_table</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">avg_log2FC</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="va">p_val_adj</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Plot cluster specific genes on a heatmap on a subset of 5000 cells.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#DefaultLayer(so[['RNA']]) &lt;- 'data'</span></span>
<span><span class="va">so.mini</span> <span class="op">&lt;-</span> <span class="va">so</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">so</span><span class="op">)</span>, size <span class="op">=</span><span class="fl">5000</span>, replace<span class="op">=</span><span class="cn">F</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">so.mini</span>, assay<span class="op">=</span><span class="st">'RNA'</span>, features <span class="op">=</span> <span class="va">marker_table_top</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> </span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-40-1.png" width="672"></div>
</div>
<div id="cell-typing-by-reference" class="section level2" number="12.2">
<h2>
<span class="header-section-number">12.2</span> Cell typing by reference<a class="anchor" aria-label="anchor" href="#cell-typing-by-reference"><i class="fas fa-link"></i></a>
</h2>
<p>There are a number of tools that classify cells into types on the basis of similarity with a reference dataset. SingleR is one such tool.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SingleR-inc/SingleR">SingleR</a></span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
## 'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
<p>But first we need to aquire a reference dataset.</p>
<p>A typical reference would be from a public single cell datasets from a matched celltype with a trusted annotation. Look for a relevant paper which has provided their GEO accession number or similar. Note that not every study remembers to publish their annotations alongside their raw data, but the authors might be willing to supply you with a processed dataset if you ask.</p>
<p>Alternativly, there are some databases of single cell datasets - see <a href="https://data.humancellatlas.org/">Human Cell Atlas</a> or <a href="https://cellxgene.cziscience.com/datasets">CELLxGENE datasets</a>. These databases also let you browse the data online - which is an excellent resource for the design or cell annotation stages; you can look up gene expression in a particular celltype to see if it matches what you are seeing.</p>
<p>There is also a package ‘celldex’ which hosts a small number of reference datasets ready to go. So for today, we will use the ‘human primary cell atalas’ reference, which includes celltypes from various different sites; Note that a sample specific reference would be better!</p>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
## 'DelayedArray::makeNindexFromArrayViewport' when loading 'HDF5Array'</code></pre>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Gives us a 'SummarizedExperiment object'</span></span>
<span><span class="va">se_ref</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/HumanPrimaryCellAtlasData.html">HumanPrimaryCellAtlasData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>A reference dataset approach can only choose from the celltypes it has.</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">se_ref</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span></code></pre></div>
<pre><code>##  [1] "DC"                   "Smooth_muscle_cells"  "Epithelial_cells"    
##  [4] "B_cell"               "Neutrophils"          "T_cells"             
##  [7] "Monocyte"             "Erythroblast"         "BM &amp; Prog."          
## [10] "Endothelial_cells"    "Gametocytes"          "Neurons"             
## [13] "Keratinocytes"        "HSC_-G-CSF"           "Macrophage"          
## [16] "NK_cell"              "Embryonic_stem_cells" "Tissue_stem_cells"   
## [19] "Chondrocytes"         "Osteoblasts"          "BM"                  
## [22] "Platelets"            "Fibroblasts"          "iPS_cells"           
## [25] "Hepatocytes"          "MSC"                  "Neuroepithelial_cell"
## [28] "Astrocyte"            "HSC_CD34+"            "CMP"                 
## [31] "GMP"                  "MEP"                  "Myelocyte"           
## [34] "Pre-B_cell_CD34-"     "Pro-B_cell_CD34+"     "Pro-Myelocyte"</code></pre>
<p>Note that singleR and celldex are bioconductor packages - they don’t natively work with Seurat objects. Rather SpatialExperiments/SpatialFeatureExperiment/SingleCellExperiments/SummarizedExperiment (For software folk; these inherit from SummarizedExperiment)</p>
<p>We could build a ‘SingleCellExperiment’ object - but singleR works just fine with a matrix of expression.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">norm_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">so</span>, assay <span class="op">=</span> <span class="st">'RNA'</span>, layer <span class="op">=</span> <span class="st">'data'</span><span class="op">)</span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu">SingleR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span><span class="op">(</span>test <span class="op">=</span> <span class="va">norm_matrix</span>, ref <span class="op">=</span> <span class="va">se_ref</span>, labels <span class="op">=</span> <span class="va">se_ref</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span></code></pre></div>
<p>Store back in the object, and plot. There are too many celltypes to show nicely, and some aren’t that relevant.</p>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span><span class="op">$</span><span class="va">singleR_labels</span> <span class="op">&lt;-</span> <span class="va">predictions</span><span class="op">$</span><span class="va">labels</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">so</span>, group.by<span class="op">=</span><span class="st">'singleR_labels'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-45-1.png" width="672"></div>
<p>TODO: Ideally use a better reference that doesn’t need this much post processing to be useful! Or make clearer what is being done here.</p>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so</span><span class="op">$</span><span class="va">singleR_delta.next</span> <span class="op">&lt;-</span> <span class="va">predictions</span><span class="op">$</span><span class="va">delta.next</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">so</span>, features<span class="op">=</span><span class="st">'singleR_delta.next'</span>, group.by<span class="op">=</span><span class="st">'singleR_labels'</span>, pt.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-46-1.png" width="672"></div>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">min_delta_next_threshold</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">min_cells_in_group</span>       <span class="op">&lt;-</span> <span class="fl">50</span></span>
<span><span class="va">so</span><span class="op">$</span><span class="va">singleR_labels_thresh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span> <span class="op">(</span> <span class="va">so</span><span class="op">$</span><span class="va">singleR_delta.next</span> <span class="op">&gt;</span> <span class="va">min_delta_next_threshold</span> ,  <span class="va">so</span><span class="op">$</span><span class="va">singleR_labels</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">so</span>, group.by<span class="op">=</span><span class="st">'singleR_labels_thresh'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-46-2.png" width="672"></div>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_per_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">so</span><span class="op">$</span><span class="va">singleR_labels_thresh</span><span class="op">)</span></span>
<span><span class="va">keep_groups</span>      <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">counts_per_group</span><span class="op">)</span><span class="op">[</span><span class="va">counts_per_group</span> <span class="op">&gt;</span> <span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">so</span><span class="op">$</span><span class="va">singleR_labels_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span> <span class="va">so</span><span class="op">$</span><span class="va">singleR_labels_thresh</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">keep_groups</span> ,    <span class="va">so</span><span class="op">$</span><span class="va">singleR_labels_thresh</span>, <span class="cn">NA</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">so</span>, group.by<span class="op">=</span><span class="st">'singleR_labels_filt'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-46-3.png" width="672"></div>
<p>Its nice to line those classifications up with clusters.</p>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Celltype proportions</span></span>
<span><span class="va">celltype_summary_table</span><span class="op">&lt;-</span> <span class="va">so</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">seurat_clusters</span>, <span class="va">singleR_labels_filt</span> <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n_cells <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## `summarise()` has grouped output by 'seurat_clusters'. You can override using the
## `.groups` argument.</code></pre>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">celltype_summary_table</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">seurat_clusters</span>, y<span class="op">=</span><span class="va">n_cells</span>, fill<span class="op">=</span><span class="va">singleR_labels_filt</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"fill"</span>, stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span> <span class="st">"Celltyping vs clustering"</span><span class="op">)</span> </span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-47-1.png" width="672"></div>
</div>
<div id="todo-cell-typeing-by-appropriate-reference" class="section level2" number="12.3">
<h2>
<span class="header-section-number">12.3</span> TODO: Cell typeing by appropriate reference<a class="anchor" aria-label="anchor" href="#todo-cell-typeing-by-appropriate-reference"><i class="fas fa-link"></i></a>
</h2>
<!-- https://cellxgene.cziscience.com/e/7357cee7-9f7f-4ab0-8cec-90de8f047e38.cxg/ -->
<!-- https://cellxgene.cziscience.com/collections/e5f58829-1a66-40b5-a624-9046778e74f5 -->
<p>Check out the annotations for large intestinve vs the cellXgene webseite: <a href="https://cellxgene.cziscience.com/e/7357cee7-9f7f-4ab0-8cec-90de8f047e38.cxg/" class="uri">https://cellxgene.cziscience.com/e/7357cee7-9f7f-4ab0-8cec-90de8f047e38.cxg/</a></p>
<p>Broad cell class looks pretty good. There’s colon included. Some individual level effect in the UMAP (stem cells limited to 1-2 samples)</p>
<p>Grab the .h5ad file for the large intestine
<a href="https://datasets.cellxgene.cziscience.com/82e3b450-6704-43de-8036-af1838daa7df.h5ad" class="uri">https://datasets.cellxgene.cziscience.com/82e3b450-6704-43de-8036-af1838daa7df.h5ad</a></p>
<p>Its h5ad format. That’s from python.</p>
<p>We can read that with schard package.</p>
</div>
<div id="spatial-examination-of-plots" class="section level2" number="12.4">
<h2>
<span class="header-section-number">12.4</span> Spatial examination of plots<a class="anchor" aria-label="anchor" href="#spatial-examination-of-plots"><i class="fas fa-link"></i></a>
</h2>
<p>The clusters were defined purely on transcriptional similarity, but when plotted on tissue, their location pattern emerges.
Here we can see that cluster 1 is epithelial cells.</p>
<p>Plotting both whole sample, and a zoomed in region of just one of the FOVs (in the cosmx definition) within it.</p>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_sample</span> <span class="op">&lt;-</span> <span class="va">so</span><span class="op">[</span>, <span class="va">so</span><span class="op">$</span><span class="va">tissue_sample</span><span class="op">==</span><span class="st">"HC_a"</span><span class="op">]</span></span></code></pre></div>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects</code></pre>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html">ImageDimPlot</a></span><span class="op">(</span><span class="va">so_sample</span>,</span>
<span>             fov          <span class="op">=</span> <span class="st">"GSM7473682.HC.a"</span>,</span>
<span>             axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             border.color <span class="op">=</span> <span class="cn">NA</span>, border.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>             group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>             boundaries   <span class="op">=</span> <span class="st">"segmentation"</span>,</span>
<span>             crop<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>             nmols <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-48-1.png" width="672"></div>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_fov</span>    <span class="op">&lt;-</span> <span class="va">so_sample</span><span class="op">[</span>, <span class="va">so_sample</span><span class="op">$</span><span class="va">fov</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects</code></pre>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html">ImageDimPlot</a></span><span class="op">(</span><span class="va">so_fov</span>,</span>
<span>             fov          <span class="op">=</span> <span class="st">"GSM7473682.HC.a"</span>,</span>
<span>             axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             border.color <span class="op">=</span> <span class="cn">NA</span>, border.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>             group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>             boundaries   <span class="op">=</span> <span class="st">"segmentation"</span>,</span>
<span>             crop<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>             nmols <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-49-1.png" width="672"></div>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html">ImageDimPlot</a></span><span class="op">(</span><span class="va">so_fov</span>,</span>
<span>             fov          <span class="op">=</span> <span class="st">"GSM7473682.HC.a"</span>,</span>
<span>             axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             border.color <span class="op">=</span> <span class="cn">NA</span>, border.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>             group.by <span class="op">=</span> <span class="st">"singleR_labels_filt"</span>,</span>
<span>             boundaries   <span class="op">=</span> <span class="st">"segmentation"</span>,</span>
<span>             crop<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>             nmols <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning: Removing 1669 cells missing data for vars requested</code></pre>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-49-2.png" width="672"></div>
</div>
<div id="recording-celltype-annotations." class="section level2" number="12.5">
<h2>
<span class="header-section-number">12.5</span> Recording celltype annotations.<a class="anchor" aria-label="anchor" href="#recording-celltype-annotations."><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Apply some cluster names</span></span>
<span><span class="va">so</span><span class="op">$</span><span class="va">cluster_code</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"c"</span>, <span class="va">so</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span>,   levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'c'</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">so</span><span class="op">$</span><span class="va">seurat_clusters</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">so</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">so</span><span class="op">$</span><span class="va">cluster_code</span></span>
<span></span>
<span><span class="va">cluster_content</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  c0 <span class="op">=</span> <span class="st">"Unknown"</span>,</span>
<span>  c1 <span class="op">=</span> <span class="st">"IntestinalEpithelia"</span>, <span class="co"># PIGR expression, epithelial location</span></span>
<span>  c2 <span class="op">=</span> <span class="st">"Fibroblasts"</span>,         <span class="co"># Collagen rich</span></span>
<span>  c3 <span class="op">=</span> <span class="st">"Plasma"</span>,              <span class="co"># MZB1, DERL3, JCHAIN</span></span>
<span>  c4 <span class="op">=</span> <span class="st">"Endothelia"</span>,          <span class="co"># Mostly endothelial in cell typing.</span></span>
<span>  c5 <span class="op">=</span> <span class="st">"Unknown"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># c5 =&gt; c5: Mono</span></span>
<span><span class="va">so</span><span class="op">$</span><span class="va">cluster_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span> <span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster_content</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">so</span><span class="op">$</span><span class="va">cluster_code</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, <span class="st">": "</span>, <span class="va">cluster_content</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">so</span><span class="op">$</span><span class="va">cluster_code</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> ,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster_content</span><span class="op">)</span>, <span class="st">": "</span>, <span class="va">cluster_content</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">so</span>, group.by<span class="op">=</span><span class="st">'cluster_labels'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-51-1.png" width="672"></div>
<div id="pull-in-real-annotation" class="section level3" number="12.5.1">
<h3>
<span class="header-section-number">12.5.1</span> Pull in real annotation<a class="anchor" aria-label="anchor" href="#pull-in-real-annotation"><i class="fas fa-link"></i></a>
</h3>
<p>The authors of this study have shared their actual celltypes. So in lieu of trying to figur this out - directly import their annotation</p>
<div class="sourceCode" id="cb135"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="st">'data/GSE234713_CosMx_annotation.csv.gz'</span></span>
<span><span class="va">anno_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span><span class="va">annotation_file</span>, show_col_types <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">anno_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">anno_table</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">anno_table</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">anno_table</span><span class="op">$</span><span class="va">id</span></span>
<span></span>
<span><span class="va">so</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">anno_table</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">so</span><span class="op">)</span>,<span class="op">]</span><span class="op">$</span><span class="va">subset</span><span class="op">)</span></span></code></pre></div>
<p>From there can use celltype labels on visualiation:</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">so_fov</span>    <span class="op">&lt;-</span> <span class="va">so</span><span class="op">[</span>, <span class="va">so</span><span class="op">$</span><span class="va">tissue_sample</span><span class="op">==</span><span class="st">"HC_a"</span> <span class="op">&amp;</span> <span class="va">so_sample</span><span class="op">$</span><span class="va">fov</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code>## Warning in so$tissue_sample == "HC_a" &amp; so_sample$fov == 1: longer object
## length is not a multiple of shorter object length</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects
## Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Centroids objects</code></pre>
<pre><code>## Warning: Not validating FOV objects
## Not validating FOV objects
## Not validating FOV objects</code></pre>
<pre><code>## Warning: Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects
## Not validating Seurat objects</code></pre>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/ImageDimPlot.html">ImageDimPlot</a></span><span class="op">(</span><span class="va">so_fov</span>,</span>
<span>             fov          <span class="op">=</span> <span class="st">"GSM7473682.HC.a"</span>,</span>
<span>             axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>             border.color <span class="op">=</span> <span class="cn">NA</span>, border.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>             group.by <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>             boundaries   <span class="op">=</span> <span class="st">"segmentation"</span>,</span>
<span>             crop<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>             nmols <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<pre><code>## Warning: Removing 1 cells missing data for vars requested</code></pre>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-53-1.png" width="672"></div>
<p>Or use them in other comparisons.</p>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Celltype proportions</span></span>
<span><span class="va">celltype_summary_table</span><span class="op">&lt;-</span> <span class="va">so</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">condition</span>, <span class="va">tissue_sample</span>, <span class="va">celltype</span> <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n_cells <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>## `summarise()` has grouped output by 'condition', 'tissue_sample'. You can override using
## the `.groups` argument.</code></pre>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">celltype_summary_table</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">tissue_sample</span>, y<span class="op">=</span><span class="va">n_cells</span>, fill<span class="op">=</span><span class="va">celltype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>position<span class="op">=</span><span class="st">"fill"</span>, stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span> <span class="st">"Celltype composition"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, ncol <span class="op">=</span> <span class="fl">1</span>, scales <span class="op">=</span> <span class="st">'free_y'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="IntroSpatialTranscriptomics_files/figure-html/unnamed-chunk-54-1.png" width="672"></div>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="clustering.html"><span class="header-section-number">11</span> Clustering</a></div>
<div class="next"><a href="analysis-example.html"><span class="header-section-number">13</span> Analysis example</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clustering-labeling"><span class="header-section-number">12</span> Clustering Labeling</a></li>
<li><a class="nav-link" href="#cluster-markers"><span class="header-section-number">12.1</span> Cluster Markers</a></li>
<li><a class="nav-link" href="#cell-typing-by-reference"><span class="header-section-number">12.2</span> Cell typing by reference</a></li>
<li><a class="nav-link" href="#todo-cell-typeing-by-appropriate-reference"><span class="header-section-number">12.3</span> TODO: Cell typeing by appropriate reference</a></li>
<li><a class="nav-link" href="#spatial-examination-of-plots"><span class="header-section-number">12.4</span> Spatial examination of plots</a></li>
<li>
<a class="nav-link" href="#recording-celltype-annotations."><span class="header-section-number">12.5</span> Recording celltype annotations.</a><ul class="nav navbar-nav"><li><a class="nav-link" href="#pull-in-real-annotation"><span class="header-section-number">12.5.1</span> Pull in real annotation</a></li></ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Intro to Spatial Transcriptomics</strong>" was written by . It was last built on 2025-09-26.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
